src/btree/bt_cursor.c

```diff
@@ -1051,17 +1052,12 @@ __wt_btcur_search_near(WT_CURSOR_BTREE *cbt, int *exactp)
         exact = cbt->compare;
         WT_ERR(__cursor_kv_return(cbt, cbt->upd_value));
         /*
-         * This compare is needed for bounded cursors in the event that a valid key is found. The
-         * returned value of exact must reflect the comparison between the found key and the
-         * original search key, not the repositioned bounds key. This comparison ensures that is the
-         * case.
+         * The returned value of exact must reflect the comparison between the found key and the
+         * original search key, not the repositioned bounds key.
          */
-        if (WT_CURSOR_BOUNDS_SET(cursor)) {
-            if (btree->type == BTREE_ROW)
-                WT_ERR(__wt_compare(session, btree->collator, &cursor->key, &state.key, &exact));
-            else
-                exact = cbt->recno < state.recno ? -1 : cbt->recno == state.recno ? 0 : 1;
-        }
+        if (WT_CURSOR_BOUNDS_SET(cursor) && bounds_reposition_exact != 0)
+            exact = bounds_reposition_exact;
+
```

Comment:
For testing purposes I suggest the following:
```
        if (WT_CURSOR_BOUNDS_SET(cursor) && bounds_reposition_exact != 0)
            exact = bounds_reposition_exact;
#ifdef HAVE_DIAGNOSTIC
        if (WT_CURSOR_BOUNDS_SET(cursor)) {
            if (btree->type == BTREE_ROW)
                WT_ERR(__wt_compare(session, btree->collator, &cursor->key, &state.key, &exact));
            else
                diagnostic_exact = cbt->recno < state.recno ? -1 : cbt->recno == state.recno ? 0 : 1;
        }
        WT_ASSERT(session, diagnostic_exact == exact);
#endif
```
This does mean that in theory we are still exposed to the heap corruption bug in diagnostic mode but it does help validate the change. We could just run this code through our python tests and then not commit it.

Reply:
I thought about this and in the long run we'd be better off without the diagnostic code, did you run stress testing with this diagnostic block in?

Reply:
Yes I did run the patch build with the diagnostic block in, no failures related this change.

Reply:
Great lets take it out then and then i'll do a final review.